### PR TITLE
SQL: Prevent StackOverflowError when parsing large statements

### PR DESCRIPTION
**Pull request by matriv**
Catch StackOverflowError exception and return a descriptive message
to the client. This prevents large statement from killing the cluster.

Fixes: #32942

### CODE REVIEW

**Comment by nik9000** [comment]
> Catch StackOverflowError

I'm not sure that this is a safe thing to do.

**Comment by matriv** [comment]
@nik9000 With this code we return an error message to the client, and the ES node is not affected.
I've searched for any other way of handling it with ANTLR4 but haven't found such. Any suggestions?

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/parser/SqlParser.java [hunk 1]
```diff
@@ -80,10 +81,13 @@ public Expression createExpression(String expression, List<SqlTypedParamValue> p
             log.debug("Parsing as expression: {}", expression);
         }
 
-        return invokeParser(expression, params, SqlBaseParser::singleExpression, AstBuilder::expression);
+        return invokeParser("expression", expression, params, SqlBaseParser::singleExpression, AstBuilder::expression);
     }
 
-    private <T> T invokeParser(String sql, List<SqlTypedParamValue> params, Function<SqlBaseParser, ParserRuleContext> parseFunction,
+    private <T> T invokeParser(String name,
```
Comment (astefan) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/parser/SqlParser.java]: Can you name the `name` parameter in a more descriptive way? As I see it, it's only used for error reporting purposes, to adapt the message to what type of entity failed to be parsed. Maybe `querySourceType`?

**Comment by nik9000** [comment]
> Catch StackOverflowError

I'm not sure that this is a safe thing to do.

> the ES node is not affected.

That is the part I'm not sure about. My instinct is that stuff like stack overflows and out of memory put the application into a weird state. I'm happy to be convinced I'm wrong about it though.

I know we've had issues in the past with out of memory leaving ES in a half functional state. So we removed catching it. And we lumped stack overflow in with that initiative.

When this has come up in other places we've mostly worked around it by untwisting the recursion into loops. If this is deep into ANTLR's code we really can't do much about it though. So I'm not sure what to do.

**Comment by romseygeek** [comment]
> When this has come up in other places we've mostly worked around it by untwisting the recursion into loops

I've run across this problem writing antlr grammars elsewhere, and the general way to solve it in my experience is to switch from recursive matching to multi-matching.  So instead of:
```
expression: part | part OR expression | ...
```
you do
```
expression: orexpression | ....
orexpression: part OR part (OR part)*
```

**Comment by jdconrad** [comment]
This problem cannot be solved using ANTLR as it requires recursion for anything more than an extremely simple grammar.  There is no way around this since expressions can be of an indeterminate length.  I view the problem as what's worse - letting a node die because of a SQL query or at least attempting to recover and catching a SOE.  The trade off of catching an SOE seems better here.

**Comment by rjernst** [comment]
Also note that stack overflow in this cannot have any negative side effects. Normally the side effects to be aware of are partial initialization of objects. Here we are constructing and using an AstBuilder. This object is the only one that could be partially initialized, but in the catch case we do not save it.

**Comment by matriv** [comment]
Please check the approach here: https://github.com/elastic/elasticsearch/pull/33902/commits/ff67d02a4f9de0b06db97991a809aff4c4327aa7

**Comment by matriv** [comment]
retest this please

**Comment by matriv** [comment]
@elasticmachine retest this please

**Comment by matriv** [comment]
My hesitation with this CircuitBreaker approach is that we cannot reset the counter for each part of the tree. This means that if we limit the elements to 100 for example, then in the query: `SELECT a=b OR a=b OR... FROM t WHERE a=b AND ...`, noElements in the SELECT clause + noElements in the WHERE clause must be <= 100.

**Comment by nik9000** [comment]
If @rjernst is ok with catching StackOverflowError here then I am as well, so long as we leave a big comment making it clear about why it is ok here. It'd be nice to link to some documentation about how StackOverflowError works and what guarantees it gives us.

**Comment by matriv** [comment]
@nik9000 Check the new implementation: 7e09985f73ce3d8d61b99c81b5ba02cf6dc1e657 (@costin 's idea)

**Comment by jdconrad** [comment]
There are no guarantees the circuit breaker presented here won't cause an SOF.  Each stack frame can have a varying size based on the input expression.  The stack frame size can also be controlled via JVM parameter so there's no guarantee of being able to calculate based on the limited number of recursions.

**Comment by matriv** [comment]
> There are no guarantees the circuit breaker presented here won't cause an SOF. Each stack frame can have a varying size based on the input expression. The stack frame size can also be controlled via JVM parameter so there's no guarantee of being able to calculate based on the limited number of recursions.

I agree with that but I after searching and reading a bit about cases of catching the `StackOverflowError`, I'm not sure we can have guarantees that JVM can continue to function without problems:

https://stackoverflow.com/questions/28551767/is-it-safe-to-catch-stackoverflowerror-in-java
https://softwareengineering.stackexchange.com/questions/209099/is-it-ever-okay-to-catch-stackoverflowerror-in-java
https://stackoverflow.com/questions/22128485/why-is-it-possible-to-recover-from-a-stackoverflowerror
https://docs.oracle.com/javase/specs/jvms/se8/html/jvms-2.html#jvms-2.5.2
https://docs.oracle.com/javase/specs/jvms/se8/html/jvms-6.html#jvms-6.3

According to the latter: `This specification cannot predict where internal errors or resource limitations may be encountered and does not mandate precisely when they can be reported. Thus, any of the VirtualMachineError subclasses defined below may be thrown at any time during the operation of the Java Virtual Machine:`

**Comment by matriv** [comment]
@elasticmachine retest this please.

**Comment by matriv** [comment]
@elasticmachine retest this please.

**Comment by matriv** [comment]
@elasticmachine retest this please.

**Comment by costin** [comment]
LGTM

**Comment by matriv** [comment]
@nik9000 @jdconrad @rjernst So, what do you think about the 2 solutions (catching StackOverflowError vs the CircuitBreakerListener) ?

**Review by astefan — APPROVED** [review]
LGTM

**Review by costin — APPROVED** [review]
LGTM

**Comment by matriv** [comment]
Backported to `6.x` with 6a4e8418ee811ba9c061b37faef1cfe88d705274

**Comment by rjernst** [comment]
@matriv Sorry I was on vacation and could not respond. I think catching StackOverflowError is the correct thing to do. We can prove based on what a stack overflow means that the state of the system has not been corrupted. We know this because we do not hang onto any objects that were being created when the SOE occurred (which is just the AstBuilder), and there are no statics that were potentially being modified.

Additionally, the commit message for this PR is deceptive because an SOE *can* still occur, it is just less likely now.

**Comment by costin** [comment]
We don't have any guarantees regarding the method execution itself or it's side-effects.
Take lazy loading: it's possible that class-loading occurs during the tree creation and some of the classes have have static initializers for constants and such; we don't know whether this have been partially initialized or not.
Or JIT-ing which might occur especially in a highly recursive scenario - is the jitted code affected? does its virtual stack still hold ? It is being unwinded so there's some safety but then again, there's also an `Error` being thrown.
If I recall correctly, in HotSpot methods share the stack frames with C/C++ native code and the VM itself (it's also JNI calls can throw SOE) so I would expect that invariants simply no longer hold and shutting things down is the only option.

**Comment by rjernst** [comment]
> Take lazy loading: it's possible that class-loading occurs during the tree creation

This is true, but stack overflow occurs from deep recursion. Classloading would have happened at the top of the recursion tree, long before we are deep enough for stack overflow.

> Or JIT-ing which might occur especially in a highly recursive scenario

JIT happens in a separate system thread. It does not stop the world in order to compile a method to native code.

**Comment by jpountz** [comment]
Even though catching stack overflow errors is sometimes ok, the challenging part is to keep this code correct over time. Eg. unrelated refactorings might mistakenly fold other statements under the `try` statement. It makes me like the circuit breaker idea a bit better: if the threshold is reasonable enough then chances that you hit a stack overflow without hitting the circuit breaker first are very thin. @romseygeek's idea is probably worth exploring as well.

**Comment by rjernst** [comment]
> I'm not sure that applies for lazy inner classes (init-on-demand idiom) that can appear at the tail of a deep recursion; e.g WHERE A OR B OR C .... OR (1000 AND ABS(POWER(1-2, 1) > 10) - the OR can fill the stack while the nested expression at the end, that triggers the error, can cause classes that haven't been loaded before.

I can see how this could be a problem, given the current design of the sql parser. In painless, all classes are loaded up front, through the whitelist. I imagine something similar could be done in sql.

> Right however it's not the compilation of the JIT but the effect of the SOE on the native (JIT-ed) stack - I'm not aware of any guarantees on that front.

I'm not sure what you mean. The JIT'd code's stack would be structured the exact same as the original code, otherwise the jvm could never safely switch to the JIT'd method's implementation in the middle of recursive calls.

> Regarding the breaker ...

The ideas here all sound plausible (although I would avoid checking the stack trace as I think the generation is dynamic, thus incurring cost to look at it). But I don't think anything can guarantee there is no stack overflow (without essentially creating our own parallel stack overflow tracking).

@jpountz re: catching SOE
> the challenging part is to keep this code correct over time

I think this is always the challenge with any pieces of code that have inter dependencies. I can see how the sql implementation is much more complicated than painless, but in the end, the idea seems to be to convert a text representation of the sql query into something that can be sent as an elasticsearch query. Given the query should be the only output, I think we can make catching SOE work. But I don't care that much, as long as the messaging of this change does not claim to "prevent" SOE, when it only makes it less likely.

**Pull request merged**